### DISCUSSION TITLE
Security best practice

### DISCUSSION THREAD

**Discussion by TtuxX** [Jip-Hop/jailmaker#111] [Q&A]
Hi there, thank you very much for your awesome tool ! :)

I wanted to know what are the best practice and how do you get an unprivileged jail? 
Coming from Core I'd like to migrate my Truenas Core Nextcloud instance. I'd love to keep my Mariadb while customising curves in nginx ; things that TrueCharts typically does not allow and on which I'm feeling like I don't have much control anymore on Scale. 
However, not being too familiar yet with Scale, I find instructions quite unclear regarding the best practice to setup a "secure" jail that could be exposed to internet.

Could you please help me and let me know what I should/should not do to secure jails created and managed with your tool? 🙂

Many thanks in advance for your kind help!

**Comment by Jip-Hop** [comment]
I advise to not expose TrueNAS SCALE or any services running on it to the internet. Instead I recommend to use a VPN, hosted on another device, to join the network where your NAS is located and access your services (Nextcloud in this case) that way.

**Comment by Lockszmith-GH** [comment]
You can expose a single port (or both 80 and 443) into a jail running a reverse proxy, and control access through that reverse proxy.

NPM (Nginx Proxy Manager) is the most WebUI oriented solution out there, but of course you can use whatever suites your fancy.

**Comment by TtuxX** [comment]
Thank you very much to you both for your help.

I realise that my question was maybe not precise enough, as my main preocccupation is more specifically about privileges and trying to get as close as possible to the Core jails security level. I think this is especially needed if I expose a service within the jail. 

I've had a look at the Security implications of the Readme but, being unfamiliar to systemd-nspawn, I'm not sure how to properly use user Namespacing and limit capabilities in the context of Jailmaker.
Could you let me know how you'd do it?

**Reply by skittle-brau** [reply]
There's a few ways you can go about it, but the way I implement user namespacing in my nspawn containers is to add: 
`--private-users=65536:65536 --private-users-ownership=chown`

This increases all user/group IDs in the container into a corresponding high range on the host, effectively giving them the same privileges as nobody/nogroup on the host.

Root user inside the container is 0, but on the host side it's 65536 — you get that number from adding 65536 to 0. In the context of Nextcloud, all files are owned by www-data which is uid 33, so outside of the container it gets mapped to 65536+33=65569. 

I run nextcloud-aio in a nspawn container so I use ACLs on my datasets in TrueNAS SCALE to give a custom user with uid 65569 (I named mine 'jail-nextcloud' access to the files needed in Nextcloud.

**Reply by Jip-Hop** [reply]
Also have a look here: https://github.com/Jip-Hop/jailmaker/tree/main/templates/podman#rootless

**Comment by Neurrone** [comment]
Edit: please ignore this, I think I finally figured out the issue. I had the impression that traverse had more permissions than modify (since it follows modify in the dropdown), giving it modify permissions fixed it.

> I run nextcloud-aio in a nspawn container so I use ACLs on my datasets in TrueNAS SCALE to give a custom user with uid 65569 (I named mine 'jail-nextcloud' access to the files needed in Nextcloud.

@skittle-brau could I check how you did this? I've been trying to figure out how to get permissions to work for the past 3 days to no avail.

I thought the best practice was having a user per container so that each container has access to only what it needs to. Hence, I have the following data set structure:

```
podman
  data
    traefik
    Plex
    ...
  stacks
```

So each container gets its own data set. They use the NFSV4 ACL.

I created a podman user on TrueNas that maps to a non-root user in the jail that I use to run my containers. It owns the podman data set, everything works as expected there - the rootless user can see the files from the jail.

I created a traefik user on TrueNas (id 526288, maps to 2000 in the jail)  and granted access to the traefik data set only. 

![permissions for the `traefik` data set, demonstrating that the `traefik` user has execute access](https://github.com/Jip-Hop/jailmaker/assets/16471962/cae22c01-4a40-4a5b-b4fe-18878a0ae322)

However when I try accessing the data via `jlmkr shell --uid 2000 jailname`, it has no access to that folder despite the ACLs being set.

```sh
[traefik@podman traefik]$ ls -al
ls: cannot open directory '.': Permission denied
```

This is what the posix permissions look like outside the jail

```sh
root@truenas[/mnt/tank/enc/docker/data]# ls -al
total 42
drwxrwx--x 4 rootless podman 4 May  6 15:29 .
drwxrwx--x 4 rootless podman 4 May  5 21:58 ..
drwxrwx--x 4 rootless podman 7 May  6 22:53 traefik
root@truenas[/mnt/tank/enc/docker/data]# cd traefik
root@truenas[/mnt/tank/enc/docker/data/traefik]# ls -al
total 88
drwxrwx--x 4 rootless podman     7 May  6 22:53 .
drwxrwx--x 4 rootless podman     4 May  6 15:29 ..
-rw------- 1 rootless podman 52310 May  6 15:38 acme.json
-rwxrwx--- 1 rootless podman   599 May  6 15:38 default.toml
drwxrwx--- 2 rootless podman     5 May  6 15:38 dynamic
drwxrwx--- 2 rootless podman     3 May  6 15:38 shared
-rwxrwx--- 1 rootless podman  1329 May  6 15:38 traefik.toml
```

My guess is that the ACL entries are completely ignored and only the permission bits are checked.

**Reply by skittle-brau** [reply]
Glad you got it figured out. I only saw this just now, so I would've said the same thing about Traverse :)
